### PR TITLE
expression, types: Function DAY/DAYOFMONTH the NoZeroDateMode sql mode check when string cast to time

### PR DESCRIPTION
**Pull request by Patrick0308**
### What problem does this PR solve?

Issue Number: a part of #11178

Problem Summary:

### What is changed and how it works?

What's Changed: check the NoZeroDateMode sql mode check when string cast to time .

How it Works:

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test

### Release note 

- Function DAY/DAYOFMONTH modify

### CODE REVIEW

**Review by qw4990 — COMMENTED** [review]
Please add cases in https://github.com/pingcap/tidb/issues/11178 as a unit test into TiDB. @Patrick0308

**Comment by Patrick0308** [comment]
@qw4990 OK

File: expression/builtin_cast.go [hunk 1]
```diff
@@ -1257,6 +1257,9 @@ func (b *builtinCastStringAsTimeSig) evalTime(row chunk.Row) (res types.Time, is
 		// Truncate hh:mm:ss part if the type is Date.
 		res.SetCoreTime(types.FromDate(res.Year(), res.Month(), res.Day(), 0, 0, 0, 0))
 	}
+	if res.IsZero() && b.ctx.GetSessionVars().SQLMode.HasNoZeroDateMode() {
+		return types.ZeroDate, true, handleInvalidTimeError(b.ctx, types.ErrWrongValue.GenWithStackByArgs(types.DateTimeStr, res.String()))
+	}
```
Comment (time-and-fate) [inline expression/builtin_cast.go:1262]: Please add a test case to cover this.

**Comment by zz-jason** [comment]
I'm going to close this PR since it hasn't been updated for a long time. feel free to reopen if you want to continue with it. thank you for your contribution.

BTW, seems the original issue has been fixed.

**Pull request closed without merging**